test/opt/instruction_list_test.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <vector>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include "opt/instruction.h"
+#include "opt/instruction_list.h"
+
+namespace {
+
+using Instruction = spvtools::ir::Instruction;
+using InstructionList = spvtools::ir::InstructionList;
+using ::testing::ContainerEq;
+using ::testing::ElementsAre;
+using InstructionListTest = ::testing::Test;
+
+class TestInstruction : public Instruction {
+ public:
+  TestInstruction() : Instruction() {
+    created_instructions_.push_back(this);
+    static int counter = 0;
+    data = ++counter;
+  }
+
+  ~TestInstruction() { deleted_instructions_.push_back(this); }
+
+  int data;
+  static std::vector<TestInstruction*> created_instructions_;
+  static std::vector<TestInstruction*> deleted_instructions_;
```

Comment:
Consider identifying instructions by data_ instead of pointers.

Reply:
I prefer the pointers because I need to know the same memory locations are destroyed, not something that happens to have the same data.